psi4/src/psi4/libmints/linalg.h

```diff
@@ -0,0 +1,494 @@
+     *  \param[in] extra
+     * TODO Generalise to any rank
+     */
+    std::string format(const std::string extra = "") const noexcept {
+        std::ostringstream retval;
+        // Title and stuff
+        if (!label_.empty()) {
+            retval << "  ## " << label_ << " " << extra << " (Symmetry " << symmetry_ << ") ##\n" << std::endl;
+        }
+        // Blocks
+        for (size_t h = 0; h < store_.size(); ++h) {
+            retval << "  Irrep: " << h + 1 << " Shape: " << detail::print_shape(store_[h].shape()) << std::endl;
+            if (store_[h].size() == 0) {
+                retval << "    (empty)" << std::endl;
+            } else {
+                retval << xt::print_options::line_width(120) << xt::print_options::precision(14) << store_[h]
+                       << std::endl;
+            }
+            retval << std::endl;
+        }
+        return retval.str();
+    }
+
+   protected:
+    unsigned int symmetry_{0};
+    std::string label_;
+    std::array<Dimension, Rank> axes_dimpi_{};
+    std::vector<Shape> shapes_{};
+    std::vector<Storage<T, Rank>> store_{};
+};
+
+template <typename T, size_t Rank>
+using SharedTensor = std::shared_ptr<Tensor<T, Rank>>;
+
+enum class Operation { None, Transpose, TransposeConj };
+
+/*! Symmetry-blocking aware GEneralized Matrix Multiplication (GEMM)
+ *  \param[in]
+ *  \param[in]
+ *  C := alpha * op(A) * op(B) + beta * C
+ */
+template <typename T>
+SharedTensor<T, 2> gemm(Operation opA, Operation opB, double alpha, const SharedTensor<T, 2>& A,
+                        const SharedTensor<T, 2>& B, double beta, SharedTensor<T, 2>& C) {
+    std::string label = "GEMM: ";
+
+    std::vector<size_t> ax_A;
+    bool transA = (opA == Operation::Transpose || opA == Operation::TransposeConj);
+    switch (opA) {
+        case Operation::None:
+            ax_A.push_back(1);
+            label += A->label();
+            break;
+        case Operation::Transpose:
+            ax_A.push_back(0);
+            label += A->label() + "T x ";
+            break;
+        case Operation::TransposeConj:
+            ax_A.push_back(0);
+            label += A->label() + "H x ";
+            break;
+    }
+
+    std::vector<size_t> ax_B;
+    bool transB = (opB == Operation::Transpose || opB == Operation::TransposeConj);
+    switch (opB) {
+        case Operation::None:
+            ax_B.push_back(0);
+            label += B->label();
+            break;
+        case Operation::Transpose:
+            ax_B.push_back(1);
+            label += B->label() + "T";
+            break;
+        case Operation::TransposeConj:
+            ax_B.push_back(1);
+            label += B->label() + "H";
+            break;
+    }
+
+    for (size_t hA = 0; hA < A->nirrep(); ++hA) {
+        size_t hB = hA ^ (transA ? 0 : A->symmetry()) ^ (transB ? B->symmetry() : 0);
+        size_t hC = hA ^ (transA ? A->symmetry() : 0);
+
+        // FIXME Currently NOT handling transpose-conjugate case
```

Comment:
should there be a NYI error somewhere for TransposeConj?

Reply:
Sounds good! I should probably also limit scalar types to non-complex.

Reply:
I have added support for `std::complex<double>` and the function can now handle transpose conjugation. The tests issue a warning, see MolSSI/QCElemental#77